File: frontend/src/lib/api/schemas/analytics.ts
import { z } from "zod";

export const dailyMetricReadSchema = z.object({
  id: z.string().uuid(),
  video_id: z.string().uuid(),
  date: z.string(),
  views: z.number(),
  watch_time_minutes: z.string(),
  likes: z.number(),
  comments: z.number(),
  shares: z.number(),
  subscribers_gained: z.number(),
  revenue_usd: z.string(),
});
export type DailyMetricRead = z.infer<typeof dailyMetricReadSchema>;

export const retentionPointReadSchema = z.object({
  id: z.string().uuid(),
  video_id: z.string().uuid(),
  date: z.string(),
  elapsed_video_percent: z.string(),
  audience_retention_percent: z.string(),
});
export type RetentionPointRead = z.infer<typeof retentionPointReadSchema>;

export const trafficSourceReadSchema = z.object({
  id: z.string().uuid(),
  video_id: z.string().uuid(),
  date: z.string(),
  source_type: z.string(),
  views: z.number(),
  watch_time_minutes: z.string(),
});
export type TrafficSourceRead = z.infer<typeof trafficSourceReadSchema>;

export const videoAnalyticsReadSchema = z.object({
  daily_metrics: z.array(dailyMetricReadSchema),
  retention_points: z.array(retentionPointReadSchema),
  traffic_sources: z.array(trafficSourceReadSchema),
});
export type VideoAnalyticsRead = z.infer<typeof videoAnalyticsReadSchema>;

export type DailyMetricIngestRequest = {
  date: string;
  views?: number;
  watch_time_minutes?: string;
  likes?: number;
  comments?: number;
  shares?: number;
  subscribers_gained?: number;
  revenue_usd?: string;
};

export type RetentionPointIngestRequest = {
  date: string;
  elapsed_video_percent: string;
  audience_retention_percent: string;
};

export type TrafficSourceIngestRequest = {
  date: string;
  source_type: string;
  views?: number;
  watch_time_minutes?: string;
};

File: backend/src/ytforge/interfaces/api/schemas/analytics.py
from __future__ import annotations

import uuid
from datetime import date as date_
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class DailyMetricIngestRequest(BaseModel):
    date: date_
    views: int = Field(default=0, ge=0)
    watch_time_minutes: Decimal = Field(default=Decimal("0"), ge=0)
    likes: int = Field(default=0, ge=0)
    comments: int = Field(default=0, ge=0)
    shares: int = Field(default=0, ge=0)
    subscribers_gained: int = 0
    revenue_usd: Decimal = Field(default=Decimal("0"), ge=0)


class DailyMetricRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    date: date_
    views: int
    watch_time_minutes: Decimal
    likes: int
    comments: int
    shares: int
    subscribers_gained: int
    revenue_usd: Decimal


class RetentionPointIngestRequest(BaseModel):
    date: date_
    elapsed_video_percent: Decimal = Field(ge=0, le=100)
    audience_retention_percent: Decimal = Field(ge=0, le=100)


class RetentionPointRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    date: date_
    elapsed_video_percent: Decimal
    audience_retention_percent: Decimal


class TrafficSourceIngestRequest(BaseModel):
    date: date_
    source_type: str = Field(min_length=1, max_length=64)
    views: int = Field(default=0, ge=0)
    watch_time_minutes: Decimal = Field(default=Decimal("0"), ge=0)


class TrafficSourceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    video_id: uuid.UUID
    date: date_
    source_type: str
    views: int
    watch_time_minutes: Decimal


class VideoAnalyticsRead(BaseModel):
    daily_metrics: list[DailyMetricRead]
    retention_points: list[RetentionPointRead]
    traffic_sources: list[TrafficSourceRead]
